### PR TITLE
feat: introduce cached flat storage deltas

### PR DESCRIPTION
**Pull request by Longarithm**
Introduce `CachedFlatStateDelta` and store it in memory instead of larger `FlatStateDelta`. There we store only hashes of trie keys instead of full trie keys. It still allows to serve read-only queries, but greatly reduces max size of deltas per block - from ~50 MB of raw entries to ~5 MB. Full motivation can be found here: https://github.com/near/nearcore/issues/8436.

Based on https://ntietz.com/blog/rust-hashmap-overhead/, hashmap overhead is 70% per entry, so we estimate full entry size as 117.

Proof that it doesn't impact performance: TODO

## Testing

Existing tests must not fail.

### CODE REVIEW

**Review by nikurt — APPROVED** [review]

File: core/store/src/flat_state.rs [hunk 1]
```diff
@@ -405,9 +396,41 @@ impl FlatStateDelta {
...
+    }
+
+    /// Returns number of all entries.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Size of all entries. May be changed if we implement inlining of `ValueRef`s.
+    fn total_size(&self) -> u64 {
+        (self.0.len() as u64) * Self::ENTRY_SIZE
```
Comment (pugachAG) [inline core/store/src/flat_state.rs:426]: Let's try to be a bit more precise here. std `HashMap` is based on [Swiss Table](https://abseil.io/blog/20180927-swisstables). In my understanding it stores all the data in flat array + a couple of bits of "control data" per entry (that is how I understood it from [the talk](https://www.youtube.com/watch?v=ncHmEUmJZf4&t=3s)). So in our case I suggest using `self.0.capacity() * (sizeof<Key> + sizeof<Value>)` and ignore control data overhead to keep it simple.

File: core/store/src/flat_state.rs [hunk 2]
```diff
@@ -405,9 +396,41 @@ impl FlatStateDelta {
...
+impl From<FlatStateDelta> for CachedFlatStateDelta {
+    fn from(delta: FlatStateDelta) -> Self {
+        Self(delta.0.iter().map(|(key, value)| (hash(key), value.clone())).collect())
+    }
+}
+
+impl CachedFlatStateDelta {
+    /// Assumed size in bytes for the cache entry. Estimated as 69 bytes for `CryptoHash` and `Option<ValueRef>` +
+    /// 70% hashmap overhead per entry ~= 117 bytes.
+    const ENTRY_SIZE: u64 = 117;
```
Comment (pugachAG) [inline core/store/src/flat_state.rs:412]: Please avoid hardcoding such values, in my understanding rust allows compile time computations using size_of: `std::mem::size_of::< CryptoHash>() + std::mem::size_of::<Option<ValueRef>>()`

**Review by pugachAG — CHANGES_REQUESTED** [review]

File: core/store/src/flat_state.rs [hunk 3]
```diff
@@ -358,12 +353,8 @@ impl FlatStateDelta {
...
     }
 
-    fn total_size(&self) -> u64 {
-        self.0.keys().map(|key| key.len() as u64 + Self::PER_ENTRY_OVERHEAD).sum()
-    }
-
     /// Merge two deltas. Values from `other` should override values from `self`.
-    pub fn merge(&mut self, other: &Self) {
+    pub fn merge(&mut self, other: Self) {
         self.0.extend(other.0.iter().map(|(k, v)| (k.clone(), v.clone())))
```
[the remark this answers is not in the document]
Comment (Longarithm) [reply core/store/src/flat_state.rs:358]: Yeah, could be nice. But there is no way to extract values from container without `mut`, and having `mut` here feels wrong.

[the remark this answers is not in the document]
Comment (pugachAG) [reply core/store/src/flat_state.rs:358]: not sure what you mean here, wouldn't `self.0.extend(other.0.into_iter())` work?

[the remark this answers is not in the document]
Comment (Longarithm) [reply core/store/src/flat_state.rs]: Um, right.

**Review by pugachAG — APPROVED** [review]
LGTM!

**Pull request merged**